doc/SAI-Proposal-ICMP-ECHO.md

```diff
@@ -0,0 +1,396 @@
+
+The objective of this document is to propose a SAI specification to simplify the ICMP-based probe as a basic service. The service primitives offered by ICMP echo include creating, destroying, and modifying a session, considering the destination address and other parameters. In response, ICMP Echo sends a signal to its clients when the ICMP Echo session starts or stops.
+
+This specification suggests:
+- The idea of an ICMP echo object or session that can be transferred to HW/ASIC or other modules for faster link detection.
+- The introduction of SAI APIs to outline ICMP echo session properties.
+- The introduction of a switch-level notification API to monitor changes in ICMP echo session states.
+
+
+## 1.1.0 Functional requirements of ICMP echo session
+- Transmit ICMP echo requests (probes) at regular, configurable intervals.
+- Identify and report changes in probe state to the control plane application.
+- Enable applications to set customized transmit and receive intervals.
+- Allow configuration of IP encapsulation parameters.
+- Establish a unique identification for each probing session.
+- Determine whether an offloaded session should process ICMP probing packets.
+- Offer statistics on ICMP echo probe and reply activities.
+
+## 2.0 Protocol overview and ICMP echo session
+ICMP echo is simple network link connectivity test protocol, similar to that most of the L3 routing protocols. ICMP Echo session runs between pair of devices exchanging ICMP echo request and reply packets.
+
+### 2.1.0 ICMP echo request and reply packet ###
+
+```C
+    0                   1                   2                   3
+    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |     Type      |     Code      |          Checksum             |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |           Identifier          |        Sequence Number        |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |     Data ...
+   +-+-+-+-+-
+```
+
+### 2.2.0 ICMP echo payload packet ###
+
+Besides the ICMP echo request and reply packets, additional information about the ICMP echo session is essential and is transmitted within the payload.
+
+***GUID:*** Global unique identifier for the ICMP echo session to correlate probe and reply packets with a session, especially when the session is offloaded to hardware, a unique identifier is required. This distinct identifier is conveyed within the ICMP payload field.
+
+***Session Cookie:*** The value of the session cookie is communicated through the ICMP payload field. The offloading engine interprets this value to decide whether the packet should undergo processing by the ICMP echo session or not.
```

Comment:
all sessions should have the same cookie on the same device, right?

Reply:
Each session can share the same cookie, but if a device needs to communicate with multiple devices, each session can have its own unique cookie. The decision to assign cookies to sessions is implementation-specific, and the device can implement a check to enforce the desired behavior.

Reply:
resolved